# T-test

In [1]:
import numpy as np
import scipy as sp
import pandas as pd

## 1 Introduction to T-test

### 1.1 Properties

1. T-test is categorized as **Numerical vs Categorical**.
2. T-test is used to compare numerical values between two categories.
3. At most, T-test supports comparison of numerical values between **two categories only**.
4. T-test is very useful in medial domain to validate clinical trials since it can work with small sample size.

### 1.2 Assumptions for T-test

1. Population Standard Deviation $\sigma$ may or may not be available.
2. T-test is used for testing hypothesis on population parameter type - Mean.
3. Sampling distribution of sample mean is normal and its mean is available.
4. T-test is used for Numerical vs Categorical features where total number of categories are $\le 2$.
5. Each category/group in the feature forms a sample.

### 1.3 Conditions for T-test

1. Data (Sample) is of type Continuous random variate.
2. Random Sampling.
3. Observations are independent of each other.
4. When sample size $n < 30$, Mean Sampling distribution should be strictly normal.
5. When sample size $n \ge 30$, CLT takes care of normality for Mean Sampling distribution.
6. Population standard deviation $\sigma$ is unknown and is estimated using Sample standard deviation $s$.
7. Homogeneity of Variance i.e., sample must have less outliers.

### 1.4 Z-Test vs T-Test

#### Why do we need T-test?

- In practice, the population standard deviation ($\sigma$) is rarely known. Z-Statistics depends on $\sigma$.
- Although the sample standard deviation ($s$) can be used to estimate $\sigma$, it introduces uncertainty because $s$ varies from sample to sample.
- T-test accounts for this uncertainty by using T-distribution instead of Standard Normal distribution when making inferences about population mean.
- When $\sigma$ is unknown, we often use a T-test instead of a Z-test.

#### Z-test vs T-test

1. As per Law of Large Numbers sample standard deviation $s$ converges to population standard deviation $\sigma$ as sample size $n$ tends to $\infty$.
2. Typically for sample size $n \ge 30$ T-test behaves like Z-test.

Sample size and population standard deviation criteria for Z-test vs T-test:

<div style="display: inline-block">

| Sample size $n$ and Population Standard Deviation $\sigma$                      | Hypothesis Test |
| :------------------------------------------------------------------------------ | :-------------- |
| Sample size $n <  30$ and population standard deviation $\sigma$ is known.      | Z-test          |
| Sample size $n <  30$ and population standard deviation $\sigma$ is NOT known.  | T-test          |
| Sample size $n \ge 30$ and population standard deviation $\sigma$ is known.     | Z-test          |
| Sample size $n \ge 30$ and population standard deviation $\sigma$ is NOT known. | Z-test / T-test |

</div>

Standard Normal Distribution vs T-Distribution: https://www.desmos.com/calculator/54fq22ixx9

### 1.5 Types of T-test

1. One Sample T-test
2. Independent Two Sample T-test
3. Paired Two sample T-test

## 2 One Sample T-test

### 2.1 Introduction

#### What is One Sample T-test?

In One Sample T-test, mean of one single sample is compared against population mean.

#### When to use One Sample T-test?

- When we do not have the value of population standard deviation.
- When the sample size is less than 30 (T-test behaves as Z-Test for sample size > 30).
- When we want to compare mean values between sample distribution and population distribution.

### 2.2 Nature of hypothesis

- $H_0: \mu = \mu_0$
- $H_a: \mu \neq \mu_0 \;(\,\text{or}\; \mu > \mu_0 \,,\, \mu < \mu_0\,)$

Where,

- $\mu$ - Variable representing Population Mean
- $\mu_0$ - Value of Population Mean

### 2.3 Test Statistic

1. The name of the test statistic in T-test is called as T-statistic.
2. Sample distribution of sample means follows T distribution.

#### Reference Distribution

Reference Distribution is **T-Distribution** under the assumption that $H_0$ is true.

#### Formula

$$
\large
\text{T-Statistic} = \frac{\bar{x} - \mu}{\displaystyle \frac{s}{\sqrt{n}}}
$$

Where,

- $\bar{x}$ - Observed mean value (Sample mean)
- $\mu$  - Population mean
- $s$ - Sample Standard Deviation
- $n$ - Sample size

### 2.4 API

```python
import scipy as sp


# Given,
sample: list = []  # Sample array
mu: float =        # Population mean

t_stat, p_value = sp.stats.ttest_1samp(
    a=sample,
    popmean=mu,
    alternative="two-sided",  # Use "less" for Left Tailed test or "greater" for Right Tailed test.
)
```

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html

## 3 Independent Two Sample T-test

### 3.1 Introduction

#### What is Independent Two Sample T-test?

- In Two Sample T-test, mean of one sample is compared with mean of second sample.
- The samples are **independent**.

#### What are Independent Samples?

- Two or more distinct, unrelated groups of data-points where the selection of one group has no influence on another.
- Both the samples are selected simultaneously at any given time (i.e., not before and after some treatment).

#### When to use Two Sample T-test?

- When we do not have the value of population standard deviation.
- When the sample size is less than 30 (T-test behaves as Z-Test for sample size > 30).
- When we want to compare mean values between independent samples from same or different population.

### 3.2 Nature of hypothesis

- $H_0: \mu_1 = \mu_2$
- $H_a: \mu_1 \neq \mu_2 \;(\,\text{or}\,\mu_1 > \mu_2 \,,\, \mu_1 < \mu_2\,)$

or,

- $H_0: \mu_1 - \mu_2 \simeq 0$
- $H_a: \mu_1 - \mu_2 \ne 0 \;(\,\text{or}\; \mu_1 - \mu_2 > 0 \,,\, \mu_1 - \mu_2 < 0\,)$

Where,

- $\mu_1$ - Mean value of Sample #1.
- $\mu_2$ - Mean value of Sample #2.

### 3.3 Test Statistic

1. The name of the test statistic used in Independent Two Sample T-test is called as T-statistic.
2. Sample distribution of sample means follows T-Distribution.

#### Formula

##### Case #1

Formula to calculate $\text{T-Statistic}$ under the assumption that **two samples have different variances**.

$$
\large
\newcommand{\mudiff}{\displaystyle \bar{x}_1 \,-\, \bar{x}_2}
\newcommand{\stderr}{\displaystyle \sqrt{\frac{s_1^2}{n_1} \,+\, \frac{s_2^2}{n_2}}}
\text{T-Statistic} = \frac{\mudiff}{\stderr}
$$

Where,

- $\bar{x}_1$ - Mean value of Sample #1.
- $\bar{x}_2$ - Mean value of Sample #2.
- $s_1$ - Sample Standard deviation value of Sample #1.
- $s_2$ - Sample Standard deviation value of Sample #2.
- $n_1$ - Sample size of Sample #1.
- $n_2$ - Sample size of Sample #2.

###### Degree of Freedom (Welch–Satterthwaite)

$$
\large
\newcommand{\sasvm}{\displaystyle \frac{s_1^2}{n_1}}
\newcommand{\sbsvm}{\displaystyle \frac{s_2^2}{n_2}}
\newcommand{\a}{\displaystyle \frac{1}{n_1 - 1}\,\biggl(\sasvm\biggr)^2}
\newcommand{\b}{\displaystyle \frac{1}{n_2 - 1}\,\biggl(\sbsvm\biggr)^2}
\newcommand{\numrtr}{{\Biggl(\sasvm\;+\;\sbsvm\Biggr)}^2}
\newcommand{\denrtr}{\a\;+\;\b}
\mathrm{df} = \frac{\numrtr}{\denrtr}
$$

### 3.4 API

```python
import scipy as sp


# Given,
sample_a: list = []  # Independent Sample 1
sample_b: list = []  # Independent Sample 2

t_stat, p_value = sp.stats.ttest_ind(
    a=sample_a, 
    b=sample_b, 
    alternative="two-sided"  # Use "less" for Left Tailed test or "greater" for Right Tailed test.
)
```

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

## 4 Paired T-test

### 4.1 Introduction

#### What is Paired T-test?

- In Paired T-test, indirectly, mean of a sample is compared with itself after undergoing some treatment.
- Mean of the difference observed in sample before and after undergoing some treatment.

#### When to use Paired T-test?

- When we do not have the value of population standard deviation.
- When the sample size is less than 30 (T-test behaves as Z-Test for sample size > 30).
- When we want to measure the impact of a treatment on sample.

### 4.2 Nature of hypothesis

- $H_0: \mu_b = \mu_a$
- $H_a: \mu_b \neq \mu_a \;(\,\text{or}\,\mu_b > \mu_a \,,\, \mu_b < \mu_a\,)$

or,

- $H_0: \mu_b - \mu_a \simeq 0$
- $H_a: \mu_b - \mu_a \ne 0 \;(\,\text{or}\; \mu_b - \mu_a > 0 \,,\, \mu_b - \mu_a < 0\,)$

Where,

- $\mu_b$ - Mean value before some treatment on Sample #1.
- $\mu_a$ - Mean value after some treatment on Sample #1.

### 4.3 Test Statistic

1. The name of the test statistic used in Independent Two Sample T-test is called as T-statistic.
2. Sample distribution of sample means follows T-Distribution.

#### Formula

$$
\large
\newcommand{numrtr}{\displaystyle \bar{x}_{\text{diff}}}
\newcommand{denrtr}{\displaystyle \frac{s}{\sqrt{n}}}
t = \frac{\numrtr}{\denrtr}
$$

Where,

- $\bar{x}_{\text{diff}}$ - Mean of the difference between sample values before and after some treatment
- $s$ - Sample standard deviation
- $n$ - Sample size

### 4.4 API

```python
import scipy as sp


# Given,
sample_a: list = []  # Sample after treatement.
sample_b: list = []  # Sample before treatement.

t_stat, p_value = sp.stats.ttest_rel(
    a=sample_a, 
    b=sample_b, 
    alternative="two-sided"  # Use "less" for Left Tailed test or "greater" for Right Tailed test.
)
```

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html

> **Note**:
>
> 1. Order of passing the samples to API does not matter but set the `alternative` parameter correctly.
> 2. Passing `alternative = "less"` to API behaves as: mu_sample_a < mu_sample_b.
> 3. Passing `alternative = "greater"` to API behaves as: mu_sample_a > mu_sample_b.

## 5 Exercises

### Exercise #1

Let's say you are a Research Scientist working on a new cognitive enhancement pill:

- The goal is to develop a pill that can significantly improve IQ scores in individuals.
- The researchers believe that the new pill will lead to a significant increase in average IQ scores for the population.

In [2]:
iq_scores = [110, 105, 98, 102, 99, 104, 115, 95]
print("Sample size:", len(iq_scores))

Sample size: 8


#### Solution

In [3]:
# STEP #1:
# H_0: mu <= 100  Given IQ scores are closer to global average.
# H_a: mu > 100   Given IQ scores are larger than global average.

In [4]:
# STEP #2:
# Distribution: T distribution
# Test statistic: T-score
# Significance level: 0.05
alpha = 0.05

In [5]:
# STEP #3:
# Since H_a has > symbol, perform right-tailed test.

Next lets calculate test-statistic, p-value, etc. without using a library.

In [6]:
# Given that average IQ of human beings are considered as 100.
mu = 100  # Population mean.
n = len(iq_scores)

In [7]:
x_bar = np.mean(iq_scores)  # Sample mean
x_bar.item()

103.5

In [8]:
s = np.std(iq_scores, ddof=1)
s.round(6).item()

6.568322

In [9]:
n = len(iq_scores)
se = s / np.sqrt(n)
se.round(6).item()

2.322253

In [10]:
# STEP #4:
t_stat = (x_bar - mu) / se
t_stat.round(6).item()

1.507157

In [11]:
p_value = 1 - sp.stats.t.cdf(t_stat, df=n - 1)
p_value.round(6).item()

0.08775

Repeat Step #4 using library.

In [12]:
# STEP #4:
t_stat, p_value = sp.stats.ttest_1samp(a=iq_scores, popmean=mu, alternative="greater")
print("t-statistic:", t_stat.round(6).item())
print("p-value", p_value.round(6).item())

t-statistic: 1.507157
p-value 0.08775


In [13]:
# STEP #5:
# Compare p-value with alpha.
if p_value <= alpha:
    print("Reject Null Hypothesis. Pill significantly improves IQ.")
else:
    print("Failed to reject Null Hypothesis. Pill has no effect on IQ.")

Failed to reject Null Hypothesis. Pill has no effect on IQ.


Average IQ of sample, 103, is **NOT** statistically significant, hence given IQ scores are closer to global average.

##### Critical Value

In [14]:
t_critical = sp.stats.t.ppf(1 - (alpha / 2), df=n - 1)
t_critical.round(6).item()

2.364624

In [15]:
# x = mu + (t_score * se)
x = mu + (t_critical * se)
x.round(0).item()

105.0

If average IQ of sample were to be 105, then the pill can be declared effective at improving IQ level.

### Exercise #2

Suppose we have IQ data samples across 2 schools, and we want to compare and see which school's students have better IQ.

- Use $\alpha$ = 0.05

In [16]:
iq_df = pd.read_csv("../0_data/01_students/iq_two_schools.csv")
iq_df.head(3)

,School,iq
0,school_1,91
1,school_1,95
2,school_1,110


In [17]:
iq_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   School  50 non-null     str  
 1   iq      50 non-null     int64
dtypes: int64(1), str(1)
memory usage: 932.0 bytes


In [18]:
iq_df["School"].value_counts()

School
school_1    26
school_2    24
Name: count, dtype: int64

In [19]:
mask = iq_df["School"] == "school_1"
sample_1 = iq_df[mask]["iq"].to_numpy()
sample_1[:5]

array([ 91,  95, 110, 112, 115])

In [20]:
mask = iq_df["School"] == "school_2"
sample_2 = iq_df[mask]["iq"].to_numpy()
sample_2[:5]

array([112, 115,  95,  92,  91])

In [21]:
x_1 = sample_1.mean()  # Sample mean of Sample #1.
x_2 = sample_2.mean()  # Sample mean of Sample #2.

print("Sample means:")
x_1.round(6).item(), x_2.round(6).item()

Sample means:


(101.153846, 109.416667)

In [22]:
s_1 = sample_1.std(ddof=1)  # Sample standard deviation of Sample #1.
s_2 = sample_2.std(ddof=1)  # Sample standard deviation of Sample #2.

print("Sample standard deviations:")
s_1.round(6).item(), s_2.round(6).item()

Sample standard deviations:


(10.924074, 13.325088)

In [23]:
n_1 = sample_1.size  # Sample size of Sample #1.
n_2 = sample_2.size  # Sample size of Sample #2.

print("Sample sizes:")
n_1, n_2

Sample sizes:


(26, 24)

##### Observations

1. Dataset contains details of:
   - 26 students from School 1 (Sample #1).
   - 24 students from School 2 (Sample #2).
2. Sample means, Sample standard deviation and Sample size details are extracted from the given dataset.
3. Two samples are of different size.

#### Solution

In [24]:
# STEP #1:
# H_0: mu_1 = mu_2  # Average IQ of students from both the schools are same.
# H_a: mu_1 != mu_2  # Average IQ of students from both the schools are significantly different.

In [25]:
# STEP #2:
alpha = 0.05

In [26]:
# STEP #3:
# Since H_a has != symbol, perform two-tailed test.

Next lets calculate test-statistic, p-value, etc. without using a library.

In [27]:
# Calculate degree of freedom.
evar_smpl1 = (s_1**2) / n_1  # Estimated variance of sample mean.
evar_smpl2 = (s_2**2) / n_2  # Estimated variance of Sample mean.
a = (evar_smpl1**2) / (n_1 - 1)
b = (evar_smpl2**2) / (n_2 - 1)

dof = ((evar_smpl1 + evar_smpl2) ** 2) / (a + b)
dof.round(6).item()

44.598367

In [28]:
# Calculate Standard Error.
se = np.sqrt(evar_smpl1 + evar_smpl2)
se.round(6).item()

3.462379

In [29]:
# STEP #4:
# Calculate t-statistic and p-value.
t_stat = (x_1 - x_2) / se
t_stat.round(6).item()

-2.386457

In [30]:
p_value = 2 * sp.stats.t.cdf(-np.abs(t_stat), df=dof)
p_value.round(6).item()

0.021317

Repeat Step #4 using library.

In [31]:
# STEP #4:
# Compute test statistic and p-value.
t_stat, p_value = sp.stats.ttest_ind(a=sample_1, b=sample_2, alternative="two-sided")
print("t-statistic:", t_stat.round(6).item())
print("p-value", p_value.round(6).item())

t-statistic: -2.405647
p-value 0.020046


In [32]:
# STEP #5:
# Compare p-value with alpha.
if p_value <= alpha:
    print("Reject Null hypothesis.")
    print("Average IQ of students from both the schools are significantly different.")
else:
    print("Failed to reject Null hypothesis.")
    print("Average IQ of students from both the schools are same.")

Reject Null hypothesis.
Average IQ of students from both the schools are significantly different.


##### Right Tailed test

In [33]:
# STEP #1:
# H_0: mu_1 = mu_2  # Average IQ of students from School 1 is similar to School 2.
# H_a: mu_1 > mu_2  # Average IQ of students from School 1 is greater than School 2.

# STEP #2:
alpha = 0.05

# STEP #3:
# Since H_a has > symbol, perform right-tailed test.

# STEP #4:
# Compute test statistic and p-value.
t_stat, p_value = sp.stats.ttest_ind(a=sample_1, b=sample_2, alternative="greater")
print("t-statistic:", t_stat.round(6).item())
print("p-value", p_value.round(6).item())

# STEP #5:
# Compare p-value with alpha.
if p_value <= alpha:
    print("Reject Null hypothesis.")
    print("Average IQ of students from School 1 is less than School 2.")
else:
    print("Failed to reject Null hypothesis.")
    print("Average IQ of students from both the schools are same.")

t-statistic: -2.405647
p-value 0.989977
Failed to reject Null hypothesis.
Average IQ of students from both the schools are same.


### Exercise #3

To analyze the importance of a revision class, students are made to attempt the same test before and after attending a revision class.  
Their scores are recorded from both the attempts in the below Dataset `problem_solving.csv`.  
Perform Paired T test to analyze the impact of the revision class.

In [34]:
ps_df = pd.read_csv("../0_data/01_students/problem_solving.csv")
ps_df.head(3)

,id,test_1,test_2
0,0,40,38
1,1,49,44
2,2,65,69


In [35]:
ps_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 137 entries, 0 to 136
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   id      137 non-null    int64
 1   test_1  137 non-null    int64
 2   test_2  137 non-null    int64
dtypes: int64(3)
memory usage: 3.3 KB


In [36]:
n, _ = ps_df.shape
print("Sample size:", n)

Sample size: 137


In [37]:
sample_before = ps_df["test_1"].to_numpy()

print("First 5 values from the sample before test attempt:")
sample_before[:5]

First 5 values from the sample before test attempt:


array([40, 49, 65, 59, 44])

In [38]:
sample_after = ps_df["test_2"].to_numpy()

print("First 5 values from the sample after test attempt:")
sample_after[:5]

First 5 values from the sample after test attempt:


array([38, 44, 69, 63, 43])

#### Solution

Perform two tailed test to see if there are any difference in test scores.

In [39]:
# STEP #1:
# H_0: mu_before = mu_after  # Revision class has no effect on scores.
# H_a: mu_before != mu_after  # Revision class has some effect on scores.

In [40]:
# STEP #2:
# Significance level
alpha = 0.05

In [41]:
# STEP #3:
# Since H_a has != symbol, perform two-tailed test.

Next lets calculate test-statistic, p-value, etc. without using a library.

In [42]:
x_diff = sample_after - sample_before
x_bar_diff = x_diff.mean()
x_bar_diff.round(6).item()

1.941606

In [43]:
s = x_diff.std(ddof=1)
s.round(6).item()

4.129817

In [44]:
se = s / np.sqrt(n)
se.round(6).item()

0.352834

In [45]:
t_stat = x_bar_diff / se
t_stat.round(6).item()

5.502886

In [46]:
p_value = 2 * sp.stats.t.cdf(-np.abs(t_stat), df=n - 1)
p_value.item()

1.7958403537922967e-07

Repeat Step #4 using library.

In [47]:
# STEP #4:
# Compute test statistic and p-value.
t_stat, p_value = sp.stats.ttest_rel(a=sample_before, b=sample_after, alternative="two-sided")
print("t-statistic:", t_stat.round(6).item())
print("p-value", p_value.item())

t-statistic: -5.502886
p-value 1.7958403537923086e-07


In [48]:
# STEP #5:
# Compare p-value with alpha.
if p_value <= alpha:
    print("Reject Null hypothesis. Revision class has some effect on scores.")
else:
    print("Failed to reject Null hypothesis. Revision class has no effect on scores.")

Reject Null hypothesis. Revision class has some effect on scores.


Two tailed test confirms that there is some significant impact of treatment on the sample.  
To identify if the impact is positive or negative lets perform Left/Right tailed test.

##### Right tailed test

Perform right tailed test to see if there are any improvements in test scores.

In [49]:
# STEP #1:
# H_0: mu_after <= mu_before  # Revision class has no effect on scores.
# H_a: mu_after > mu_before  # Revision class has significantly improved scores.

# STEP #2:
# Significance level
alpha = 0.05

# STEP #3:
# Since H_a has > symbol, perform right-tailed test.

# STEP #4:
# Compute test statistic and p-value.
t_stat, p_value = sp.stats.ttest_rel(a=sample_after, b=sample_before, alternative="greater")
print("t-statistic:", t_stat.round(6).item())
print("p-value", p_value.item())

# STEP #5:
# Compare p-value with alpha.
if p_value <= alpha:
    print("Reject Null hypothesis. Revision class has significantly improved scores.")
else:
    print("Failed to reject Null hypothesis. Revision class has no effect on scores.")

t-statistic: 5.502886
p-value 8.979201768961543e-08
Reject Null hypothesis. Revision class has significantly improved scores.
